src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs

```diff
@@ -513,23 +513,115 @@ public virtual TestProcessStartInfo GetTestHostProcessStartInfo(
+                    // won't reach this code, but let's write this in a generic way anyway, to avoid breaking if we add more variants of testhost*.exe.
+                    var environmentVariableName = $"DOTNET_ROOT_{_architecture.ToString().ToUpperInvariant()}";
+
+                    var existingDotnetRoot = _environmentVariableHelper.GetEnvironmentVariable(environmentVariableName);
+                    if (!StringUtilities.IsNullOrWhiteSpace(existingDotnetRoot))
+                    {
+                        // The variable is already set in the surrounding environment, don't set it, because we want to keep what user provided.
+                    }
+                    else
+                    {
+                        // Set the architecture specific variable to the environment of the process so it is picked up.
+                        startInfo.EnvironmentVariables.Add(environmentVariableName, dotnetRootPath);
+                    }
+                }
+                else
+                {
+                    // This is an old testhost that built against netcoreapp3.1, it does not understand architecture specific DOTNET_ROOT_<ARCH>, we have to set it more carefully
+                    // to avoid setting DOTNET_ROOT that points to x64 but is picked up by x86 host.
+                    //
+                    // Also avoid setting it if we are already getting it from the surrounding environment.
+                    var architectureFromEnv = (Architecture)Enum.Parse(typeof(Architecture), dotnetRootArchitecture, ignoreCase: true);
+                    if (architectureFromEnv == _architecture)
+                    {
+                        if (_architecture == Architecture.X86)
+                        {
+                            const string dotnetRootX86 = "DOTNET_ROOT(x86)";
+                            if (StringUtils.IsNullOrWhiteSpace(_environmentVariableHelper.GetEnvironmentVariable(dotnetRootX86)))
+                            {
+                                startInfo.EnvironmentVariables.Add(dotnetRootX86, dotnetRootPath);
+                            }
+                        }
+                        else
+                        {
+                            const string dotnetRoot = "DOTNET_ROOT";
+                            if (StringUtils.IsNullOrWhiteSpace(_environmentVariableHelper.GetEnvironmentVariable(dotnetRoot)))
+                            {
+                                startInfo.EnvironmentVariables.Add(dotnetRoot, dotnetRootPath);
+                            }
+                        }
+                    }
+                }
             }
             else
             {
-                EqtTrace.Verbose($"DotnetTestHostmanager.LaunchTestHostAsync: Prefix '{prefix}*' not found in env variables");
+                // Fallback, can delete this once the change is in dotnet sdk. because they are always used together.
```

Comment:
Is SDK our only concern here? Don't we insert in Test Explorer as well? Does Test Explorer do anything with `VSTEST_WINAPPHOST_*`?

Reply:
it is only sdk.